payment/vote.go

```diff
@@ -5,84 +5,260 @@ import (
+	Channel       string          `json:"channel" valid:"-"`
+	ID            uuid.UUID       `json:"id"`
+	CreatedAt     time.Time       `json:"createdAt"`
+	BaseVoteValue decimal.Decimal `json:"baseVoteValue"`
+	VoteTally     int64           `json:"voteTally"`
+	FundingSource string          `json:"fundingSource"`
+}
+
+// NewVoteEvent - Create a new VoteEvent given a Vote
+func NewVoteEvent(v Vote) (*VoteEvent, error) {
+	var (
+		ve = &VoteEvent{
+			ID:            uuid.NewV4(),
+			Type:          v.Type,
+			Channel:       v.Channel,
+			CreatedAt:     time.Now().UTC(),
+			VoteTally:     v.VoteTally,
+			FundingSource: "uphold",
+		}
+		err error
+	)
+	// default base vote value
+	if ve.BaseVoteValue, err = decimal.NewFromString("0.25"); err != nil {
+		return nil, fmt.Errorf("failed to default BaseVoteValue: %w", err)
 	}
 
-	_, err = govalidator.ValidateStruct(vote)
+	return ve, nil
+}
+
+// CodecEncode - encode using avro vote codec
+func (ve *VoteEvent) CodecEncode(codec *goavro.Codec) ([]byte, error) {
+	return codec.BinaryFromNative(nil, map[string]interface{}{
+		"type":          ve.Type,
+		"channel":       ve.Channel,
+		"id":            ve.ID.String(),
+		"createdAt":     ve.CreatedAt.Format(time.RFC3339),
+		"baseVoteValue": ve.BaseVoteValue.String(),
+		"voteTally":     ve.VoteTally,
+		"fundingSource": ve.FundingSource,
+	})
+}
+
+// CodecDecode - Decode using avro vote codec
+func (ve *VoteEvent) CodecDecode(codec *goavro.Codec, binary []byte) error {
+	native, _, err := codec.NativeFromBinary(binary)
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to decode avro payload: %w", err)
 	}
 
-	requestCredentials := make([]cbr.CredentialRedemption, len(credentials))
-	issuers := make(map[string]*Issuer)
+	// gross
+	v, err := json.Marshal(native)
+	if err != nil {
+		return fmt.Errorf("unable to marshal avro payload: %w", err)
+	}
 
-	for i := 0; i < len(credentials); i++ {
-		var ok bool
-		var issuer *Issuer
+	err = json.Unmarshal(v, ve)
+	if err != nil {
+		return fmt.Errorf("unable to decode decoded avro payload to VoteEvent: %w", err)
+	}
 
-		publicKey := credentials[i].PublicKey
+	return nil
+}
 
-		if issuer, ok = issuers[publicKey]; !ok {
-			issuer, err = service.datastore.GetIssuerByPublicKey(publicKey)
-			if err != nil {
-				return errors.Wrap(err, "Error finding issuer")
+// DrainVoteQueue - event loop that waits a certain amount of time, and then
+// grabs from the vote queue and applies redemptions.
+func (service *Service) DrainVoteQueue(ctx context.Context, cadence time.Duration) {
+	go func(ctx context.Context) {
+	OUTER:
+		for {
+			select {
+			case <-ctx.Done():
+				// cancellation happened, kill this worker
+				log.Printf("cancellation envoked in drain vote queue!\n")
+				return
+			case <-time.After(cadence):
+				// pull vote from db queue
+				tx, records, err := service.datastore.GetUncommittedVotesForUpdate(ctx)
+				if err != nil {
+					if tx != nil {
```

Comment:
It looks like we're re-using the same rollback/logging in each section - Do you think we should extract this into a function?

Reply:
:+1: